subgraphs/compound-v3/src/sdk/account.ts

```diff
@@ -0,0 +1,331 @@
+import {
+  BigDecimal,
+  BigInt,
+  Bytes,
+  ethereum,
+  log,
+} from "@graphprotocol/graph-ts";
+import {
+  Account,
+  LendingProtocol,
+  Market,
+  Position,
+  PositionSnapshot,
+  PositionCounter,
+} from "../../generated/schema";
+import {
+  BIGINT_ZERO,
+  exponentToBigDecimal,
+  INT_ONE,
+  INT_ZERO,
+  TransactionType,
+} from "./constants";
+import { TokenManager } from "./token";
+
+/**
+ * This file contains the AccountClass, which does
+ * the operations on the Account entity. This includes:
+ *  - Creating a new Account
+ *  - Updating an existing Account
+ *  - Making a position
+ *  - Making position snapshots
+ *
+ * Schema Version: 3.0.0
+ * Last Updated: Dec 4, 2022
+ * Author(s):
+ *  - @dmelotik
+ */
+export class AccountManager {
+  private isNew: boolean; // true if the account was created
+  private account: Account;
+  private positonError: boolean;
+  private position!: Position;
+  private market: Market;
+  private protocol: LendingProtocol;
+  private event: ethereum.Event;
+
+  constructor(
+    account: Bytes,
+    market: Market,
+    protocol: LendingProtocol,
+    event: ethereum.Event
+  ) {
+    this.market = market;
+    this.protocol = protocol;
+    this.event = event;
+    let _account = Account.load(account);
+    if (!_account) {
+      _account = new Account(account);
+      _account.positionCount = INT_ZERO;
+      _account.openPositionCount = INT_ZERO;
+      _account.closedPositionCount = INT_ZERO;
+      _account.depositCount = INT_ZERO;
+      _account.withdrawCount = INT_ZERO;
+      _account.borrowCount = INT_ZERO;
+      _account.repayCount = INT_ZERO;
+      _account.liquidateCount = INT_ZERO;
+      _account.liquidationCount = INT_ZERO;
+      _account.transferredCount = INT_ZERO;
+      _account.receivedCount = INT_ZERO;
+      _account.flashloanCount = INT_ZERO;
+      _account.save();
+      this.isNew = true;
+    } else {
+      this.isNew = false;
+    }
+    this.account = _account;
+  }
+
+  // returns true if the account was created in this instance
+  isNewAccount(): boolean {
+    return this.isNew;
+  }
+
+  addPosition(
+    asset: Bytes,
+    newBalance: BigInt,
+    side: string,
+    transactionType: string,
+    priceUSD: BigDecimal,
```

Comment:
Not a big deal, but useful nevertheless. Let the library have access to pricing. Make it depend on an interface
```
interface Pricer {
    getAmountPriceUSD(amount: BigInt, token: Token): BigDecimal
}

class AccountManager {
    ...
    prices: Pricer;
}
```
this way you can omit the price argument from everywhere.

Reply:
I see what you are doing here. What would you expect to happen if the pricing function implementation can't get the price? $0 or an error?

Reply:
Hmm, that’s a good point. Do we ever abort some process if we cannot fetch a price?

To keep things simple I would return `0` and work with that. If we realise it is necessary to sometimes abort when prices cannot be fetched we can add that functionality. Most times if the price lib returns 0 we just go with that value AFAIK, so it shouldn’t matter.

If when building a subgraph we wanted to not create a position if unable to fetch the price, we could make it return a `class { price: BigDecimal, error: boolean}` instead.

Another option is exposing both methods, one where you can pass the price and one which uses the pricer. But I think this can get confusing for a user of the lib.

 What do you think?

Reply:
Yeah, I think 0 would be fine. I like the ideal of having a price interface that we can use in all subgraphs and then implement the class specific to each protocol

Reply:
tbh @jaimehgb it doesn't necessarily make sense in this case. For example, I get prices bc I have a market loaded. I have no need to get prices for any arbitrary token.

Reply:
I’m not sure I understand. But I think it still applies. Maybe one particular instance of `AccountManager` will be dealing with a single `Market` at any given time, but the code itself will deal with multiple depending on what market is instantiated. 

If it will always be just 1 token price per instance you could pass the price of the token at the time of the instantiation instead of a `Pricer` interface though.

Reply:
Plan: see how it affects other subgraphs and iterate from there